Expose access to tensor stride

Mikolaj-A-Kowalski opened the pull request:
I believe it will be useful feature to be able to access the strides of a tensor from Fortran. I feel like it should help with debugging/understanding what the `_from_array` tensor constructors are doing.  

In addition I have removed the `UNIX` preprocessor variable and associated `ifdef-else`. I *think* it is possible to avoid them by using appropriate typedefs. I separated it into its own commit so it would be easier to drop if I am missing something here ;-)  It is certainly something to discuss if we want integrated.  

I have marked it as draft until we decide if we want to integrate these features.

### TODO
- [x] Update  `CHANGELOG.md` (?)

jatkinson1000 reviewed (commented):
Thanks @Mikolaj-A-Kowalski can confirm it works on Mac!
It's great to remove some jank, so thanks for updating this!
How did you find that they were using `int64_t`?

Would be good to test on Windows, but that can become part of reinstating the Windows CI I think.

Couple of comments/queries added inline.
Could also consider adding to the "tensor interrogation" example - I also want to make an example around these aspects as part of #415 where this will be very useful.

The other thing is I think you'll need to update `FTorch.md` to remove the `UNIX` macro.

Mikolaj-A-Kowalski commented on src/ctorch.h:
```diff
@@ -25,6 +25,14 @@ typedef void *torch_int_t;
+// We can as well to avoid portability problems
+// (i.e. 64bit integer beeing long long on Windows)
+//
+typedef int64_t torch_size_t;
```
I have went with this name by analogy with `std::size_t`, which is generally used for all indexing- and size-related computations for the  standard containers. So I think that the name is ok in that sense. 

But I think the comment may be misleading. Will change to make the above more explicit.

Mikolaj-A-Kowalski commented:
> How did you find that they were using int64_t?

I didn't go too deep, but just looked at the return types of the [`stride`](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/core/TensorBase.h#L181) (and other) methods for the A10 BaseTensor. For all the size/stride-related functionality they use  `int64_t` or the `c10::SymInt` which claims to have ["API equivalent to int64_t"](https://github.com/pytorch/pytorch/blob/main/c10/core/SymInt.h#L21).  

> Couple of comments/queries added inline.
Could also consider adding to the "tensor interrogation" example - I also want to make an example around these aspects as part of https://github.com/Cambridge-ICCS/FTorch/pull/415 where this will be very useful.

> The other thing is I think you'll need to update FTorch.md to remove the UNIX macro.

:ok_hand:

Mikolaj-A-Kowalski commented:
@jatkinson1000 I think I have addressed all of your comments and I think this one should be ready for more through review and to be integrated. 

When adding the extra example I have decided to go with minimal modification and just modify `Example_1` with few lines to spit out shape and strides of the tensor. I felt that this might be the most appropriate place for that. The alternative I was thinking of was to create a 2nd executable in the `Example_1` which would focus on interrogations of the tensor.  This is what we might do if the number of things to showcase grows (e.g. as a result of #415).

jatkinson1000 commented on examples/1_Tensor/tensor_manipulation.f90:
Perhaps a note in the comments to say what the 'expected' result of these should be would be useful?

jatkinson1000 commented on src/ctorch.h:
```diff
@@ -25,6 +25,15 @@ typedef void *torch_int_t;
 // Opaque pointer type alias for float scalars
 typedef void *torch_float_t;
 
+// Type that represent size, strides and indexing on tensors
```
```suggestion
// Type that represents size, strides and indexing on tensors
```

jatkinson1000 commented on src/ctorch.h:
```diff
@@ -25,6 +25,15 @@ typedef void *torch_int_t;
+// Type that represent size, strides and indexing on tensors
+// (like std::size_t for standard containers)
+//
+// Torch is using internally int64_t [i.e. signed 64bit integer] (30/07/2025)
```
What does this date signify?

Mikolaj-A-Kowalski replied:
Wanted to make a record when I checked in case it will stop being valid. 

But one can just `git blame` if one needs. Will remove.

jatkinson1000 commented on test/unit/unittest_tensor_interrogation.pf:
```diff
@@ -156,6 +156,54 @@ contains
+    integer(kind=c_int64_t), parameter :: tensor_shape(ndims) = [6]
+    integer(kind=c_int64_t), parameter :: expected_stride(ndims) = [1]
+
+    ! Create a tensor with uninitialised values and check torch_tensor_get_shape can correctly
+    ! identify its shape
```
```suggestion
    ! Create a tensor with uninitialised values and check torch_tensor_get_stride can correctly
    ! identify its stride pattern
```

jatkinson1000 commented on test/unit/unittest_tensor_interrogation.pf:
```diff
@@ -156,6 +156,54 @@ contains
+    integer(kind=c_int64_t), parameter :: tensor_shape(ndims) = [2,3]
+    integer(kind=c_int64_t), parameter :: expected_stride(ndims) = [3, 1]
+
+    ! Create a tensor with uninitialised values and check torch_tensor_get_shape can correctly
+    ! identify its shape
```
```suggestion
    ! Create a tensor with uninitialised values and check torch_tensor_get_stride can correctly
    ! identify its stride pattern
```

jatkinson1000 commented on test/unit/unittest_tensor_interrogation.pf:
```diff
@@ -156,6 +156,54 @@ contains
+    integer(kind=c_int64_t), parameter :: tensor_shape(ndims) = [1,2,3]
+    integer(kind=c_int64_t), parameter :: expected_stride(ndims) = [6, 3, 1]
+
+    ! Create a tensor with uninitialised values and check torch_tensor_get_shape can correctly
+    ! identify its shape
```
```suggestion
    ! Create a tensor with uninitialised values and check torch_tensor_get_stride can correctly
    ! identify its stride pattern
```

jatkinson1000 commented on test/unit/unittest_tensor_interrogation.pf:
```diff
@@ -156,6 +156,54 @@ contains
+
+  end subroutine test_get_stride_2D
+
+  ! Unit test for the torch_tensor_get_shape method of a 3D tensor
```
```suggestion
  ! Unit test for the torch_tensor_get_stride method of a 3D tensor
```

jatkinson1000 commented on test/unit/unittest_tensor_interrogation.pf:
```diff
@@ -156,6 +156,54 @@ contains
+
+  end subroutine test_get_stride_1D
+
+  ! Unit test for the torch_tensor_get_shape method of a 2D tensor
```
```suggestion
  ! Unit test for the torch_tensor_get_stride method of a 2D tensor
```

jatkinson1000 commented on test/unit/unittest_tensor_interrogation.pf:
```diff
@@ -156,6 +156,54 @@ contains
 
   end subroutine test_get_shape_3D
 
+  ! Unit test for the torch_tensor_get_shape method of a 1D tensor
```
```suggestion
  ! Unit test for the torch_tensor_get_stride method of a 1D tensor
```

jatkinson1000 commented on test/unit/unittest_tensor_constructors_destructors.pf:
I don't object, but just wondered your reasoning for adding a stride and shape check here?

jatkinson1000 reviewed (changes requested):
Hi @Mikolaj-A-Kowalski Thanks for this.

I'm broadly happy with this PR now, just a couple of queries/suggestions.

Also a bunch of sneaky copy-paste typos that I'm happy for you to collapse into a single commit.

Mikolaj-A-Kowalski commented:
I think that all the comments should be addressed.

Sorry for the delay.
